walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,92 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> ret = new ArrayList<>();
+    if (request.getDuration() <= 0) {
+      return ret;
+    }
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return ret;
+    }
+    List<TimeRange> occupiedBlocks = recordOccupiedBlocks(events, request.getAttendees());
+    ret = complement(TimeRange.WHOLE_DAY, occupiedBlocks);
+    ret.removeIf(tr -> tr.duration() < request.getDuration());
+    return ret;
+  }
+
+  /* Return list of blocks occupied by attendees, unordered */
+  private List<TimeRange> recordOccupiedBlocks(Collection<Event> events, Collection<String> names) {
+    List<TimeRange> occupiedBlocks = new ArrayList<>();
+    for (Event event : events) {
+      if (isOccupied(event, names)) {
+        occupiedBlocks.add(event.getWhen());
+      }
+    }
+    return occupiedBlocks;
+  }
+
+  /* Return true if anyone in the targetNames is involved in the event */
+  private boolean isOccupied(Event event, Collection<String> targetNames) {
+    for (String targetName : targetNames) {
+      if (event.getAttendees().contains(targetName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /* Return a list of the available TimeRanges */
+  private List<TimeRange> complement(TimeRange wholeDay, List<TimeRange> mergedTimeRanges) {
+    Stack<TimeRange> complStack = new Stack<>();
+    Stack<TimeRange> rawStack = new Stack<>();
+
+    // TimeRanges are added in the order that they are returned by mergedTimeRanges's iterator
+    // The earliest (by start time) TimeRange is on top.
+    mergedTimeRanges.sort(TimeRange.ORDER_BY_START.reversed());
+    rawStack.addAll(mergedTimeRanges);
+
+    complStack.push(wholeDay);
+    while (!rawStack.isEmpty()) {
+      TimeRange tr1 = complStack.pop();
+      TimeRange tr2 = rawStack.pop();
+      // tr2.end()<=tr1.end()
+      complHelper(tr1, tr2, complStack);
+    }
+    return complStack;
   }
+
+  /* Compute the difference tr1\tr2, and push the result to the stack */
+  private void complHelper(TimeRange tr1, TimeRange tr2, Stack<TimeRange> stack) {
+    // tr2.end()<=tr1.end() guaranteed
+    if (tr1 == tr2) {
+      return;
+    } else if (!tr1.overlaps(tr2)) {
+      //          |--tr1--|
+      // |--tr2--|
+      stack.push(tr1);
+    } else if (!tr1.contains(tr2)) {
+      //    |--tr1--|
+      // |--tr2--|
```

Comment:
I believe you are missing a case. If you know tr1 and tr2 do overlap and tr1 doesn't contain tr2, you can't guarantee what you have in the drawing. It could be that tr2 fully contains tr1. 

I'm not fully sure of what solution you are implementing so maybe your code still works with missing this case. I good way to find out (before you fix this) is to write the a test case that reaches this case and tr2 contains tr1.

Reply:
`tr1.end()` should always equal to `TimeRange.END_OF_DAY` so tr2 can't fully contain tr1 (assuming tr2 is legal). One edge case is when they both end at the END_OF_DAY, in that case a zero-length timeRange will be returned and filtered out because it is too short.

    |--tr1--|
|---tr2----|

Reply:
I see. Didn't know that without knowing the theory haha. If something is always constant then there might be a way to express this solution in a clearer way.... BUT don't lol. Don't want you to spend more time on this unit.

Reply:
tooo late already modified the code before reading your comments 🤣

Reply:
Thanks for all the comments tho, I know this is a super long PR and the code wasn't quite well written. I learnt a lot from yall comments